### PR TITLE
User loses member-level access to site if membership payment has expired.

### PR DESCRIPTION
**Pull request by patmbolger**
PT Story: https://www.pivotaltracker.com/story/show/153496002

Changes proposed in this pull request:
1. Implemented logic to reflect business rule stated in title.
2. When a member logs in we check if their latest membership payment has expired.  If so, they are set to non-member status.

Screenshots (Optional):

Ready for review:
@AgileVentures/shf-project-team

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -44,11 +44,9 @@ def default_url_options
   end
 
   def after_sign_in_path_for(resource)
-    if resource.admin?
-      admin_root_path
-    else
-      information_path
-    end
+    return admin_root_path if resource.admin?
```
Comment (weedySeaDragon) [inline app/controllers/application_controller.rb:47]: why the change?  they are logically equivalent, and the if...else..  conditional is (arguably) better sofware engineering.  (IOW, putting a `return` in there is like having a GOTO statement.  ack!)

I am _not_ a fan of having `return` statements within a method unless it is absolutely unavoidable.  This method is so simple, that I don't see the need for it.

File: app/models/user.rb [hunk 2]
```diff
@@ -30,6 +30,10 @@ def membership_payment_notes
     payment_notes(Payment::PAYMENT_TYPE_MEMBER)
   end
 
+  def membership_current?
```
Comment (weedySeaDragon) [inline app/models/user.rb:33]: This makes me realize that we need to be sure to review all of the `...member...?` type methods so that we're certain that they are all synchronized.  For instance:  what is the relationship to this method (which looks at the expiration date) and the `member?` method -- which is based on the User model attribute.  Are they all going to give us consistent answers?
(nothing is wrong with this method.  Just making me think more about them all)

File: config/initializers/after_sign_in.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+Warden::Manager.after_set_user except: :fetch do |user, _auth, _opts|
+  if user&.member? && ! user.membership_current?
```
Comment (weedySeaDragon) [inline config/initializers/after_sign_in.rb]: This should be a method in the User model so that (a) the logic and functionality for updating the state of the User (the membership state, in this case) is in the model.  In the future, when we need to make this richer (perhaps do more checks or do more changes when the conditions are met), we will not be prone to forgetting that this code is tucked away in the initializers.
(Also:  this really _is_ the responsibility of the User model as we have it right now.  Although I can certainly see a future where we have a class whose responsibility it is to review and update the membership status)

File: features/edit_company.feature [hunk 4]
```diff
@@ -9,6 +9,10 @@ Feature: As a member
       | applicant_3@happymutts.com |       | false     |
       | admin@shf.se               | true  | true      |
 
+    Given the following payments exist
+      | user_email                 | start_date | expire_date | payment_type | status | hips_id |
+      | applicant_1@happymutts.com | 2017-10-1  | 2017-12-31  | member_fee   | betald | none    |
```
Comment (weedySeaDragon) [inline features/edit_company.feature]: Use a more meaningful email name so that the intent/role/purpose of the user is clear both in the scenarios and in looking at the raw data (e.g. in the db tables).

How about "paid_member@happymutts.com" ?  That will make the difference between this user & the other one obvious.  (one is paid, the other isn't)

File: spec/helpers/application_helper_spec.rb [hunk 5]
```diff
@@ -311,22 +317,58 @@
...
-      response = /class="No".*#{user_payment.expire_date}/
-      expect(expire_date_label_and_value(user)).to match response
+    context 'company' do
+
+      it 'returns date with style "yes" if expire_date more than a month away' do
+        brand_payment.update(expire_date: Time.zone.today + 2.months)
+        response = /class="Yes".*#{brand_payment.expire_date}/
+        expect(expire_date_label_and_value(company)).to match response
+      end
+
```
Comment (weedySeaDragon) [inline spec/helpers/application_helper_spec.rb]: What is supposed to happen when expire_date == 1 month away?  (either change the wording in the "it.." statements to have one of them include '=' or include another test for that specific case)

File: spec/helpers/application_helper_spec.rb [hunk 6]
```diff
@@ -311,22 +317,58 @@
...
+      it 'returns date with style "maybe" if expire_date less than a month away' do
+        brand_payment.update(expire_date: Time.zone.today + 2.days)
```
Comment (weedySeaDragon) [inline spec/helpers/application_helper_spec.rb]: your `it...` statement specifies "less than a month" but the time that you're actually setting is just +2.days
Should have the date that you're using as close as possible to what you've described in the "it..." statement.

File: app/helpers/application_helper.rb [hunk 7]
```diff
@@ -163,22 +163,26 @@ def boolean_radio_buttons_collection(text_vals = { true: 'Yes', false: 'No' })
   def expire_date_label_and_value(entity)
```
Comment (weedySeaDragon) [inline app/helpers/application_helper.rb]: This method doesn't belong in this class.   The responsibility of this class is for view-related functionality that is or can be applied to all views in the application.  This method applies only views connected to Users and Companies (and really, only to those things that have _expiration dates_).

This should be refactored out.  Commonality should be abstracted into a common class or module.
The UsersHelper and CompaniesHelper classes can then incorporate them (via `include` or whatever) as needed.

**Review by weedySeaDragon — CHANGES_REQUESTED** [review]
I haven't stood up these changes yet (but will).  here are my "surface review" thoughts.  
enjoy ;-)

File: config/locales/sv.yml [hunk 8]
```diff
@@ -484,6 +484,7 @@ sv:
       edit_branding_status: Redigera Status
       branding_license: Accepterar policy
       pay_branding_fee_tooltip: Betald avgift förlänger H-märkningen i ett år.
+      branding_fee_expire_date_tooltip: Branding-licensen löper ut efter detta datum
```
Comment (thesuss) [inline config/locales/sv.yml]: Efter detta datum går perioden för din H-märkning ut.

File: config/locales/sv.yml [hunk 9]
```diff
@@ -621,6 +622,7 @@ sv:
       submit_new_password: Spara det nya lösenordet
       member: *member
       pay_membership_tooltip: När du betalar din medlemsavgift förlängs ditt medlemskap i ett år.
+      membership_expire_date_tooltip: Medlemskapet löper ut efter detta datum
```
Comment (thesuss) [inline config/locales/sv.yml]: Efter detta datum går perioden för ditt medlemskap ut.

**Review by thesuss — CHANGES_REQUESTED** [review]

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (patmbolger) [reply app/controllers/application_controller.rb:47]: Actually, this is an example of a"guard clause", and many people prefer them over `if ...else` statements.

Ruby community standards recommends this: https://github.com/bbatsov/ruby-style-guide#no-nested-conditionals. (and thus rubocop should call this out also).

If you google "ruby guard clause" you'll find plenty of material about this technique.

Grimm ("Confident Ruby") strongly recommends this as well.

In any case, the reason I changed this is that I initially put the check for member payment status in here, then later pulled it out as it did not really belong here.

File: app/models/user.rb (hunk 2 above)
Comment (patmbolger) [reply app/models/user.rb:33]: That's a good question, and I wish we could just use _one_ attribute to determine membership.  Here, however, I can't - as payment status is changed independent of `member` status, but `member` status is set to false if payment status is expired.

Anyway, if `membership_current?` is `false`, we set`member` to `false` upon login.

So, no, we won't get consistent answers if we have an expired member fee payment, but the user has not logged in since expiration.

Longer term, we could run a job at midnight that set `member` based on payment expiration date.

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (weedySeaDragon) [reply app/controllers/application_controller.rb:47]: Okey doke.  Now that you've let me grumble, I'll go along with that style. :-)

**Comment by patmbolger** [comment]
@weedySeaDragon @thesuss - changes made.  Thanks for comments.

**Review by weedySeaDragon — APPROVED** [review]

**Review by thesuss — APPROVED** [review]
